src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettings.cs

```diff
@@ -0,0 +1,146 @@
+        {
+            _environmentSettings = environmentSettings ?? throw new ArgumentNullException(nameof(environmentSettings));
+            _globalSettingsFile = globalSettingsFile ?? throw new ArgumentNullException(nameof(globalSettingsFile));
+            _paths = new Paths(environmentSettings);
+            environmentSettings.Host.FileSystem.CreateDirectory(Path.GetDirectoryName(_globalSettingsFile));
+            _watcher = environmentSettings.Host.FileSystem.WatchFileChanges(_globalSettingsFile, FileChanged);
+        }
+
+        private void FileChanged(object sender, FileSystemEventArgs e)
+        {
+            SettingsChanged?.Invoke();
+        }
+
+        public event Action? SettingsChanged;
+
+        public async Task<IDisposable> LockAsync(CancellationToken token)
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(nameof(GlobalSettings));
+            }
+            token.ThrowIfCancellationRequested();
+            // We must use Mutex because we want to lock across different processes that might want to modify this settings file
+            var mutex = await AsyncMutex.WaitAsync($"812CA7F3-7CD8-44B4-B3F0-0159355C0BD5{_globalSettingsFile}".Replace("\\", "_").Replace("/", "_"), token, Unlocked).ConfigureAwait(false);
+            _locked = true;
+            return mutex;
+        }
+
+        private void Unlocked()
+        {
+            _locked = false;
+        }
+
+        public void Dispose()
+        {
+            if (_disposed)
+            {
+                return;
+            }
+            _disposed = true;
+
+            _watcher?.Dispose();
+            _watcher = null;
+        }
+
+        public async Task<IReadOnlyList<TemplatePackageData>> GetInstalledTemplatesPackagesAsync(CancellationToken cancellationToken)
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(nameof(GlobalSettings));
+            }
+
+            if (!_environmentSettings.Host.FileSystem.FileExists(_globalSettingsFile))
+            {
+                return Array.Empty<TemplatePackageData>();
+            }
+
+            for (int i = 0; i < 5; i++)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                try
+                {
+                    string? textFileContent = _paths.ReadAllText(_globalSettingsFile, "{}");
+                    GlobalSettingsData? data = JsonConvert.DeserializeObject<GlobalSettingsData>(textFileContent);
+                    return data.Packages ?? Array.Empty<TemplatePackageData>();
+                }
+                catch (Exception)
+                {
+                    if (i == 4)
+                    {
+                        throw;
+                    }
+                }
+                await Task.Delay(20).ConfigureAwait(false);
+            }
```

Comment:
Idea is... Its possible that other process locks this file while reading/writing and that operation should be done within ~20ms so lets just retry, but if we are already retrying... Lets try few more times... And after ~100ms of trying its time to give up and throw exception because problem is probably something else...